File: IPWorks OpenPGP Samples/Sign and Encrypt Email with Attachments/c++/pfilemailer.cpp
/*
 * IPWorks OpenPGP 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks OpenPGP in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksopenpgp
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworksopenpgp.h"

#define LINE_LEN 80
#define MESSAGE_LEN 1024

class MyKeyMgr: public KeyMgr
{
  virtual int FireKeyList(KeyMgrKeyListEventParams *e)
  {
    printf("%-58s %-8s %-12s\n", e->UserId, e->KeyId, e->HasSecretKey ? "private" : ""); 
    return 0;
  }
};

int main(int argc, char **argv)
{
	PFileMailer pfilemailer;
  MyKeyMgr keymgr;

	char buffer[LINE_LEN];     // input text buffer
  char keyringDir[LINE_LEN];
  int ret_code;

  printf("***************************************************************\n");
  printf("* This demo shows how to use the PFileMailer and KeyMgr       *\n");
  printf("* components to sign and encrypt email messages.              *\n");
  printf("***************************************************************\n\n");

	printf("Mail Server: ");
	fgets(buffer, LINE_LEN, stdin);
	buffer[strlen(buffer)-1] = '\0';
	pfilemailer.SetMailServer(buffer);

  printf("Mail Port: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer) - 1] = '\0';
  pfilemailer.SetMailPort(atoi(buffer));

  pfilemailer.SetSSLStartMode(SSL_AUTOMATIC);

	printf("From: ");
	fgets(buffer, LINE_LEN, stdin);
	buffer[strlen(buffer)-1] = '\0';
	pfilemailer.SetFrom(buffer);

	printf("To: ");
	fgets(buffer, LINE_LEN, stdin);
	buffer[strlen(buffer)-1] = '\0';
	pfilemailer.SetSendTo(buffer);

  printf("User: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer) - 1] = '\0';
  pfilemailer.SetUser(buffer);

  printf("Password: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer) - 1] = '\0';
  pfilemailer.SetPassword(buffer);

	printf("Subject: ");
	fgets(buffer, LINE_LEN, stdin);
	buffer[strlen(buffer)-1] = '\0';
	pfilemailer.SetSubject(buffer);

  printf("Enter the message. To end the message, enter \".\" on a single line by itself.\n");
	printf("Message:\n");

	char message[MESSAGE_LEN];
	message[0] = '\0';
	while(fgets(buffer, LINE_LEN, stdin))
	{
		buffer[strlen(buffer)-1] = '\0';
		strcat(message, buffer);
		if (strcmp(buffer, ".") == 0)
			break;
	}

	pfilemailer.SetMessageText(message);

	// if you want to add attachments:
	printf("Enter file names to attach. (press enter to stop adding attachments) ...\n");
	printf("File Name: ");
	while (fgets(buffer, LINE_LEN, stdin))
	{
		buffer[strlen(buffer)-1] = '\0';
		if (strlen(buffer) > 0)
		{
			pfilemailer.AddAttachment(buffer);
			printf("File Name: ");
		}
		else
			break;
	}

  printf("Please select an option: (0: Cancel, 1: Sign, 2: Encrypt, 3: Sign and Encrypt)");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  int cmd = atoi(buffer);

  if (cmd < 1 || cmd > 3) goto done;
  
  printf("Keyring Dir: ");
  fgets(keyringDir, LINE_LEN, stdin);
  keyringDir[strlen(keyringDir)-1] = '\0';
  ret_code = keymgr.LoadKeyring(keyringDir);
  if (ret_code) goto done;

  printf("\nListing Keys...\n");
  printf("%-58s %-8s %-12s\n", "User ID", "Key ID", "Private Key"); 
  keymgr.ListKeys();

  if (cmd == 0)
  {
    goto done;
  }
  else if (cmd  == 1) // sign
  { 
    pfilemailer.SetKeyCount(1);
    pfilemailer.SetKeyKeyring(0, keyringDir);

    printf("\nPrivate Key User ID (Used to sign outgoing messages): ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
    pfilemailer.SetKeyUserId(0, buffer);

    printf("Passphrase: ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
    pfilemailer.SetKeyPassphrase(0, buffer);

    ret_code = pfilemailer.Sign();
    if (ret_code) goto done;
  }
  else if (cmd == 2) // encrypt
  {    
    pfilemailer.SetRecipientKeyCount(1);
    pfilemailer.SetRecipientKeyKeyring(0, keyringDir);

    printf("Recipient Key User ID (Used to encrypt outgoing messages): ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
    pfilemailer.SetRecipientKeyUserId(0, buffer);

    ret_code = pfilemailer.Encrypt();
    if (ret_code) goto done;
  }
  else if (cmd == 3) // sign and encrypt
  {
    pfilemailer.SetKeyCount(1);
    pfilemailer.SetKeyKeyring(0, keyringDir);

    printf("\nPrivate Key User ID (Used to sign outgoing messages): ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
    pfilemailer.SetKeyUserId(0, buffer);

    printf("Passphrase: ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
    pfilemailer.SetKeyPassphrase(0, buffer);

    pfilemailer.SetRecipientKeyCount(1);
    pfilemailer.SetRecipientKeyKeyring(0, keyringDir);

    printf("Recipient Key User ID (Used to encrypt outgoing messages): ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
    pfilemailer.SetRecipientKeyUserId(0, buffer);

    ret_code = pfilemailer.SignAndEncrypt();
    if (ret_code) goto done;
  }
  else
  {
    printf("Invalid Command.\n");
    goto done;
  }

	printf("Sending message ...\n");
	ret_code = pfilemailer.Send();

	if (ret_code)     // Got an error.  The user is done.
	{
		goto done;
	}
	else
	{
		printf("Message sent successfully\n");
	}

done:
  if (ret_code)    // Got an error.  The user is done.
  {
    printf("\nError: %d", ret_code);
    if (keymgr.GetLastError())
    {
      printf(" \"%s\"\n", keymgr.GetLastError());
    }
    if (pfilemailer.GetLastError())
    {
      printf(" \"%s\"\n", pfilemailer.GetLastError());
    }
  }
	fprintf(stderr, "Exiting... (press enter)\n");
	getchar();
	exit(ret_code);
	return 0;
}

File: IPWorks OpenPGP Samples/Sign and Encrypt Email with Attachments/java/pfilemailer.java
/*
 * IPWorks OpenPGP 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks OpenPGP in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksopenpgp
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworksopenpgp.*;

public class pfilemailer extends ConsoleDemo {
  PFileMailer mailer = new PFileMailer();
  KeyMgr keymgr = new KeyMgr();
  
  boolean encrypt = false;
  boolean sign = false;
  
  public pfilemailer() {
    String keyringDir = "";

    try {
      System.out.println("**************************************************************");
      System.out.println("IPWorks OpenPGP Java Edition - PFilemailer Demo Application.");
      System.out.println("This demo shows how to use the PFilemailer component to send  ");
      System.out.println("encrypted and signed emails.                                  ");
      System.out.println("**************************************************************\n");

      keymgr.addKeyMgrEventListener(new DefaultKeyMgrEventListener() {
        public void keyList(KeyMgrKeyListEvent e) {
          System.out.println(String.format("%-80s [%-8s] %-12s", e.userId, e.keyId, (e.hasSecretKey ? "private" : "")));
        }
      });        

      while(true) {        
        int command = Character.getNumericValue(ask("Please enter a command", ":", "(0: quit, 1: configure pgp settings, 2: send new email)"));
        if (command < 0 || command > 2) {
          System.out.println("Invalid Command.");
          continue;
        }
        if (command == 0) {
          break;
        } else if(command == 1) { //Configure PGP settings.
          while(true) {
            System.out.println("Current PGP Settings...");
            System.out.println("\tKeyring:                " + 
                      (keymgr.getKeyring().length() > 0 ? keymgr.getKeyring() : "Not set"));
            System.out.println("\tPrivate Key:            " + 
                      (mailer.getKeys().size() > 0 ? mailer.getKeys().get(0).getUserId() : "None selected"));
            System.out.println("\tPrivate Key Passphrase: " + (mailer.getKeys().size() > 0 ? 
                      (mailer.getKeys().get(0).getPassphrase().length() > 0 ? "Set" : "Not set") : "Not set"));
            System.out.println("\tRecipient Key:          " + 
                      (mailer.getRecipientKeys().size() > 0 ? mailer.getRecipientKeys().get(0).getUserId() : "Non selected"));
            System.out.println();
            int option = Character.getNumericValue(ask("Please select an operation", ":", "(0: back, 1: change keyring, " + 
                      "2: list keys, 3: change private key, 4: change recipient key)"));
            
            if(option == 0) {       //Go back to previous menu.
              break;
            } else if(option == 1) {   //Specify keyring
              boolean keyringLoaded = false;
              while(!keyringLoaded) {
                try {
                  String entry = prompt("Keyring Directory (Enter \":q\" to exit)");
                  if(entry.equalsIgnoreCase(":q"))
                    break;
                  keymgr.loadKeyring(entry);
                  keyringDir = entry;
                  keyringLoaded = true;
                } catch(Exception ex) {
                  displayError(ex);      
                }
              }

              System.out.println("\nListing keys...\n");
              System.out.println(String.format("%-80s %-10s %-12s", "User ID", "Key ID", "Private Key"));
              keymgr.listKeys();
              System.out.println();
            } else if(option == 2) {   //List keys
              System.out.println("\nListing keys...\n");
              System.out.println(String.format("%-80s %-10s %-12s", "User ID", "Key ID", "Private Key"));
              keymgr.listKeys();
              System.out.println();
            } else if(option == 3) {   //Specify private key
              boolean keyLoaded = false;
              while(!keyLoaded) {
                try {
                  String entry = prompt("UserId (Enter \":q\" to exit, \":c\" to clear entry)");
                  if(entry.equalsIgnoreCase(":q")) {
                    break;
                  } else if(entry.equalsIgnoreCase(":c")) {
                    mailer.getKeys().clear();
                  } else {
                    mailer.getKeys().clear();
                    mailer.getKeys().add(new Key(keyringDir, entry));
                    mailer.getKeys().get(0).setPassphrase(prompt("Passphrase"));
                  }
                  keyLoaded = true;
                } catch(Exception ex) {
                  displayError(ex);
                }
              }              
            } else if(option == 4) {  //Specify recipient key
              boolean keyLoaded = false;
              while(!keyLoaded) {
                try {
                  String entry = prompt("Recipient UserId (Enter \":q\" to exit, \":c\" to clear entry)");
                  if(entry.equalsIgnoreCase(":q")) {
                    break;
                  } else if(entry.equalsIgnoreCase(":c")) {
                    mailer.getRecipientKeys().clear();
                  } else {
                    mailer.getRecipientKeys().clear();
                    mailer.getRecipientKeys().add(new Key(keyringDir, entry));
                  }
                  keyLoaded = true;
                } catch(Exception ex) {
                  displayError(ex);
                }
              }
            } else {          //Invalid option specified
              System.out.println("Invalid command.");
            }
          }
        } else if(command == 2) {  //Send an email
          mailer.setMailServer(prompt("Enter mail server"));
          mailer.setMailPort(Integer.parseInt(prompt("Port")));
          mailer.setFrom(prompt("From"));
          mailer.setSendTo(prompt("To"));
          mailer.setSubject(prompt("Subject"));
          
          mailer.setSSLStartMode(0); // auto

          //Use these properties for client authentication.
          mailer.setUser(prompt("User"));
          mailer.setPassword(prompt("Password"));
          
          mailer.config("ProcessAttachments=true");
          
          System.out.println("Please enter the message. When finished enter \":q\" on a line by itself:");
                BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
                String temp = "";
                String message = "";
                while (!(temp = bf.readLine()).equalsIgnoreCase(":q")) {
                  message += temp + "\n";
                }
                mailer.setMessageText(message);
          
                
                mailer.getAttachments().clear();
                System.out.println("Please enter the full path to attachments that should be sent, one per line." + 
                          "When finished, enter \":q\" on a line by itself:");
                while(!(temp = bf.readLine()).equalsIgnoreCase(":q")) {
                  mailer.getAttachments().add(new FileAttachment(temp));
                }
                
                if(mailer.getKeys().size() > 0 && mailer.getRecipientKeys().size() > 0) {
                  mailer.signAndEncrypt();
                } else if(mailer.getKeys().size() > 0) {
                  mailer.sign();
                } else if(mailer.getRecipientKeys().size() > 0) {
                  mailer.encrypt();
                }
                
                mailer.send();
                System.out.println("Message sent.");
        }
      }      
    } catch(Exception e) {
      displayError(e);
    }
  }

  public static void main(String[] args) {
    new pfilemailer();
  }

}


class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksOpenPGPException) {
      System.out.print(" (" + ((IPWorksOpenPGPException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}



